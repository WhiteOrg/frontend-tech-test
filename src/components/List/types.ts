type TList = {
  items: TItems<TItem<any>>;
  render: (item: TItem<any>, index: number) => JSX.Element;
};

type TItem<T> = T;

type TItems<T> = T[];

export type { TList, TItem, TItems };
