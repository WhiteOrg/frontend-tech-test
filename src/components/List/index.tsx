import React, { FC } from "react";

import { TList } from "./types";
import {
  StyledItemWrapper as ItemWrapper,
  StyledSeparator as Separator,
  StyledListWrapper as ListWrapper,
} from "./styles";

const List: FC<TList> = ({ items, render }) => {
  return (
    <ListWrapper>
      {items.map((item, index) => (
        <>
          <ItemWrapper>{render(item, index)}</ItemWrapper>
          <Separator />
        </>
      ))}
    </ListWrapper>
  );
};

export { List };
