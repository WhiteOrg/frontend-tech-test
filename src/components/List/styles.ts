import styled from "styled-components";
import { Flex } from "reflexbox";

const StyledSeparator = styled.div`
  height: 1px;
  background-color: #f5f4fe;
`;

const StyledItemWrapper = styled.div`
  color: #000000;
  padding: 10px 5px;
`;

const StyledListWrapper = styled(Flex)`
  flex-direction: column;
  padding: 0 15px 0 10px;
  max-height: 300px;
  overflow-y: auto;
`;

export { StyledSeparator, StyledItemWrapper, StyledListWrapper };
