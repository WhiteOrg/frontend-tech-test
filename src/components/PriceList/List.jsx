import React from 'react'
import { getDataFromAPI } from '../../utils'
import styled from 'styled-components'

const Container = styled.div`
    width: 100%;
    padding: 5px;
    margin-top: 1em;
    border-bottom: 2px solid #eee;
    padding-bottom: 1em;
    font-weight: bold;
`
const UnorderedList = styled.ul`
    width: 100%;
    height: 300px;
    overflow-y: scroll;
    position: relative;
    ::before {
        width: 100%;
        height: 30px;
    }
    &::-webkit-scrollbar {
        width: 5px;
    }
    &::-webkit-scrollbar-thumb {
        background-color: #f9cb11;
        border-radius: 10px;
    }
    &::-webkit-scrollbar-track {
        background-color: #eee;
    }
`

const ListItem = styled.li`
    max-height: 100px;
    height: 50px;
    display: flex;
    align-items: center;
    border-bottom: 1px solid #eee;

`
const Index = styled.span`
    display: block;
    width: 50px;
`
const Title = styled.span`
    display: block;
`
const Prize = styled.span`
    display: block;
    flex: 1;
    text-align: right;
    margin-right: 1.5rem;
`

//JSX
export default function List({ refreshList, setRefreshList }) {

    const [ priceItems, setPriceItems ] = React.useState([])
    const [ isLoading, setIsLoading ] = React.useState(true)

    function getList() {
        getDataFromAPI().then( data => {
            setPriceItems(data.prizeList)
            // SETTING UP TIMEOUT JUST TO PREVENT FROM FAST BLINKING LOADER ON REFRESH
            setTimeout(() => {
                setIsLoading(false)
            }, 300)
        })
    }
    // WHEN COMPONENT DID MOUNT -- RUNS ONLY ONCE AT START
    React.useEffect( () => {
        getList()
    }, [])

    // UPDATES EACH TIME THE DEPENDENCY ARRAY WILL CHANGE AND ALSO ON COMPONENT DID MOUNT
    React.useEffect( () => {
        
        if (refreshList === true) {
            setIsLoading(true)
            getList()
            setRefreshList(!refreshList)
        }
    }, [refreshList, setRefreshList])

    return(
        <Container>
            <UnorderedList>
                {
                    isLoading ? <div className="prizeLoader">Loading...</div> :
                    priceItems.map( (item, index) => (
                        <ListItem key={index}>
                            <Index>{index + 1}</Index>
                            <Title>{item.title}</Title>
                            <Prize>{item.total}</Prize>
                        </ListItem>
                    ))
                }
            </UnorderedList>
        </Container>
    )
}