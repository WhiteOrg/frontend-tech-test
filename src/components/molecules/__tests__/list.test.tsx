import React from "react";
import renderer from "react-test-renderer";
import List from "../list";

describe("MOLECULES: Render testing", () => {
  test("Should render List", () => {
    const prizeList = [
      {
        title: "P.C",
        total: "£2,500.00",
      },
      {
        title: "J.S",
        total: "£1,250.00",
      },
      {
        title: "H.J",
        total: "£700.00",
      },
      {
        title: "J.V",
        total: "£300.00",
      },
      {
        title: "L.K",
        total: "£100.00",
      },
      {
        title: "H.D",
        total: "£50.00",
      },
      {
        title: "V.G",
        total: "£35.00",
      },
      {
        title: "D.C",
        total: "£20.00",
      },
      {
        title: "P.D",
        total: "£10.00",
      },
      {
        title: "L.M",
        total: "£5.00",
      },
    ];
    const component = renderer.create(
      <div>
        <List prizeList={prizeList} />
      </div>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
