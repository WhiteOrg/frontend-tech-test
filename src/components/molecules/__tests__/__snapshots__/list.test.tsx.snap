// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ATOMS: Render testing Should render Hr 1`] = `
<div>
  <div
    className="sc-AxirZ gdibQh"
  >
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        1
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        P.C
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £2,500.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        2
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        J.S
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £1,250.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        3
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        H.J
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £700.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        4
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        J.V
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £300.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        5
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        L.K
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £100.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        6
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        H.D
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £50.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        7
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        V.G
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £35.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        8
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        D.C
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £20.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        9
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        P.D
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £10.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        10
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        L.M
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £5.00
      </p>
    </div>
  </div>
</div>
`;

exports[`ATOMS: Render testing Should render List 1`] = `
<div>
  <div
    className="sc-AxirZ gdibQh"
  >
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        1
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        P.C
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £2,500.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        2
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        J.S
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £1,250.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        3
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        H.J
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £700.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        4
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        J.V
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £300.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        5
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        L.K
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £100.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        6
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        H.D
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £50.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        7
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        V.G
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £35.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        8
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        D.C
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £20.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        9
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        P.D
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £10.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        10
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        L.M
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £5.00
      </p>
    </div>
  </div>
</div>
`;

exports[`MOLECULES: Render testing Should render List 1`] = `
<div>
  <div
    className="sc-AxirZ gdibQh"
  >
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        1
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        P.C
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £2,500.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        2
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        J.S
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £1,250.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        3
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        H.J
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £700.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        4
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        J.V
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £300.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        5
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        L.K
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £100.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        6
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        H.D
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £50.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        7
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        V.G
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £35.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        8
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        D.C
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £20.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        9
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        P.D
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £10.00
      </p>
    </div>
    <div
      className="sc-AxiKw hrigKI"
    >
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        10
      </p>
      <p
        className="sc-AxjAm DSefD title-col"
        color="black"
        size="medium"
      >
        L.M
      </p>
      <p
        className="sc-AxjAm DSefD"
        color="black"
        size="medium"
      >
        £5.00
      </p>
    </div>
  </div>
</div>
`;
